mirrord-layer/src/file/hooks.rs

```diff
@@ -0,0 +1,379 @@
+use std::{ffi::CStr, io::SeekFrom, os::unix::io::RawFd, path::PathBuf, ptr, slice};
+
+use frida_gum::{interceptor::Interceptor, Module, NativePointer};
+use libc::{self, c_char, c_int, c_void, off_t, size_t, ssize_t, AT_FDCWD, FILE};
+use mirrord_protocol::ReadFileResponse;
+use tracing::error;
+
+use super::{
+    ops::{fdopen, fopen, openat},
+    OpenOptionsInternalExt, IGNORE_FILES, OPEN_FILES,
+};
+use crate::file::ops::{lseek, open, read, write};
+
+/// Hook for `libc::open`.
+///
+/// **Bypassed** by `raw_path`s that match `IGNORE_FILES` regex.
+pub(super) unsafe extern "C" fn open_detour(raw_path: *const c_char, open_flags: c_int) -> RawFd {
```

Comment:
move logic to safe functions for all hooks please (as much as possible)

Reply:
Alright, my last time trying to argue in favor of a bit more separation:

1. Much of the code in the file `_detour` functions is either Rust `unsafe` or C unsafe;
2. Moving all the code to the `ops` module is detrimental to the safe wrapper design (imo). Ideally, the safe wrappers should only produce safe errors;
3. It'll be more work when refactoring to support _not-unix_ platforms, the safe wrappers just require a bit of _massaging_  to not deal with `RawFd`, but if the the whole function is `libc` specific, then it'll be more work;
4. The safe wrappers in `ops` would need more work to deal with the potential UB, or other weirdness, assume:

```rust
unsafe fn foo_detour() {
    // I assume `foo` handles the possible UB cases and other weirdness, otherwise it should not be safe.
    foo();
}

fn foo() { ... }
```

See this function for [NonZero::new_unchecked](https://doc.rust-lang.org/src/core/num/nonzero.rs.html#56-62) where it is marked as `unsafe`, even though there is nothing requiring it to actually be marked as such.